WebHostLib/customserver.py

```diff
@@ -271,34 +272,73 @@ def get_ssl_context():
 
     loop = asyncio.get_event_loop()
 
+    async def start_ws_server(ctx: WebHostContext, port: int) -> int | None:
+        ctx.server = websockets.serve(
+            functools.partial(server, ctx=ctx), ctx.host, port, ssl=get_ssl_context())
+        await ctx.server
+        # Get the port number from the started server instance:
+        port = None
+        for wssocket in ctx.server.ws_server.sockets:
+            socketname = wssocket.getsockname()
+            if wssocket.family == socket.AF_INET6:
+                # Prefer IPv4, as most users seem to not have working IPv6 support
+                if not port:
+                    port = socketname[1]
+            elif wssocket.family == socket.AF_INET:
+                port = socketname[1]
+        return port
+
+    async def start_room_server_rnd(ctx: WebHostContext) -> int | None:
+        attempts = 0
+        port: int | None = None
+
+        allow_fallback: bool = room_ports_config["overflow"]
+        max_attempts: int = room_ports_config["max_attempts"]
+
+        while port is None:
+            attempts += 1
+            if attempts > max_attempts:
+                if not allow_fallback:
+                    if e is None:
+                        return None
+                    else:
+                        raise e
+                else:
+                    port = 0
+            else:
+                port = get_random_port(room_ports_config)
+
+            try:
+                port = await start_ws_server(ctx, port)
+            except OSError as e:
+                if port == 0:
+                    raise
+                port = None
+        return port
+
+    async def start_room_server(ctx: WebHostContext) -> int | None:
+        # Prevent marked-as-broken rooms from trying to obtain another port.
+        if ctx.port < 0:
+            return None
+
+        try:
+            port = await start_ws_server(ctx, ctx.port)
+        except OSError as start_oserr:
+            port = await start_room_server_rnd(ctx)
+        return port
+
     async def start_room(room_id):
         with Locker(f"RoomLocker {room_id}"):
             try:
                 logger = set_up_logging(room_id)
                 ctx = WebHostContext(static_server_data, logger)
-                ctx.load(room_id)
+                ctx.load(room_id, room_ports_config)
```

Comment:
Semantically, I think it makes no sense to pass room_ports_config into `load`. I'd suggest to move it to be an attribute of the context (pass it into WebHostContext's `__init__` in the line above.